Use node addresses from informer in kubelet certificate manager

aojea opened the pull request:
The kubelet certificate manager was using a closure to get the node addresses, but this closure depended on a static field that was only updated during the node status update. This created a twisted dependency between the node.status reconcile loops and the certificate manager.

This commit fixes this issue by using the node informer to get the node addresses directly. This ensures that the kubelet always requests a certificate with the latest node addresses.

/kind bug
/kind cleanup

Fixes #130001

```release-note
NONE
```

/sig node
/sig network

aojea commented:
/assign @liggitt @dims 
/cc @lentzi90 @mengqiy

dims commented:
cc @cartermckinnon @mmerkes @tzneal

adrianmoisey commented on pkg/kubelet/certificate/kubelet.go:
```diff
@@ -70,9 +72,32 @@ func newGetTemplateFn(nodeName types.NodeName, getAddresses func() []v1.NodeAddr
+		klog.ErrorS(err, "fail to obtain node from local cache", "node", nodeName)
+		return nil
+	}
+	if len(node.Status.Addresses) == 0 {
```
This is my lack of experience with the Kubernetes codebase, is it possible that `node.Status` is nil at this stage?

adrianmoisey replied:
Looking at other parts of the code base, it seems to be safe to assume that `node.Status` isn't nil

aojea replied:
if no node is found we get an err NotFound IIRC, so at this point we should have a node object

adrianmoisey replied:
The part I was wondering about, is then a node is returned but doesn't yet contain status, that way `node.Status` would be nil, and `node.Status.Addresses` would return a nil pointer exception.

But I see that this is safe (I think I was confused with pointer references), since this can't happen.

lentzi90 reviewed (commented):
This is such a nice improvement! Thanks!
LGTM from my side

liggitt commented on pkg/kubelet/certificate/kubelet.go:
```diff
@@ -70,9 +72,32 @@ func newGetTemplateFn(nodeName types.NodeName, getAddresses func() []v1.NodeAddr
+
 // NewKubeletServerCertificateManager creates a certificate manager for the kubelet when retrieving a server certificate
 // or returns an error.
-func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg *kubeletconfig.KubeletConfiguration, nodeName types.NodeName, getAddresses func() []v1.NodeAddress, certDirectory string) (certificate.Manager, error) {
```
you shouldn't need to change this file, just change the function passed in to get node addresses, right?

aojea committed:
kubelet: Use node addresses from informer

The kubelet certificate manager was using a closure to get the node addresses,
but this closure depended on a static field that was only updated during the node
status update. This created a twisted dependency between the node.status
reconcile loops and the certificate manager.

This commit fixes this issue by using the node informer to get the node addresses directly.
This ensures that the kubelet always requests a certificate with the latest node addresses.

aojea commented on pkg/kubelet/certificate/kubelet.go:
```diff
@@ -70,9 +72,32 @@ func newGetTemplateFn(nodeName types.NodeName, getAddresses func() []v1.NodeAddr
+		return nil
+	}
+	// wait for the cloud provider to initialize the node addresses
+	for _, taint := range node.Spec.Taints {
```
@liggitt updated

cartermckinnon commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -466,3 +467,28 @@ func (kl *Kubelet) setCachedMachineInfo(info *cadvisorapiv1.MachineInfo) {
+	// wait for the cloud provider to initialize the node addresses
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+			return nil
```
So kubelet will no longer request a cert until the taint is removed? I think that makes sense, given the external cloud provider "owns" the address list, but this will lengthen the period when kubelet is ready/accepting pods without having a serving cert. Can you add a log line here?

Somewhat related to the discussion in: https://github.com/kubernetes/kubernetes/pull/73047

aojea replied:
yeah, related to this https://github.com/kubernetes/kubernetes/pull/125813

The node addresses are updated just after the taint is removed when the node is initialized by the cloud controller manager

https://github.com/kubernetes/kubernetes/blob/8cca6d98a4fb855979d2a0642391b7a441673d78/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L463-L472

Bear in mind that the CSR requires to have at least one IP , that mean there should be one IP in node.Status.Addresses

https://github.com/kubernetes/kubernetes/blob/8cca6d98a4fb855979d2a0642391b7a441673d78/pkg/kubelet/certificate/kubelet.go#L50-L51

But I'm not sure of what path we want to follow:

- no external cloud provider, everything is the same
- external cloud provider
   -  no --node-ip specified, effectively the behavior should be the same, as there will be no IPs on node.status.addresses
   - node-ip specified, this is the tricky part per https://github.com/kubernetes/kubernetes/pull/121028
       - ip matches cloud-provider addresses, only one CSR is sent and we'll regress on node startup
       - ip partially matches, two CSR are sent, but it may regress or not
       - ip does not match, two CSR are sent but only the last one will be the one matching the actual addresses.

We have been so many times back and forth with node.status.address that now I think we should remove this check, and keep the existing behavior, if someone is specifies the --node-ip with the flag, then it may want to send the CSR with that IP ...

aojea commented:
Pushed last version based on discussion with @cartermckinnon in https://github.com/kubernetes/kubernetes/pull/130362#discussion_r1973629179, this keeps the same behavior, just removes the dependency on the node status loops by using the local informer. Based on the last year experience _better the devil you know than the devil you don't_

aojea commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -466,3 +466,13 @@ func (kl *Kubelet) setCachedMachineInfo(info *cadvisorapiv1.MachineInfo) {
+
+// getLastStableNodeAddresses returns the last observed node addresses.
+func (kl *Kubelet) getLastObservedNodeAddresses() []v1.NodeAddress {
+	node, err := kl.GetNode()
```
let's be defensive in case GetNode() changes and returns nil in the future in some cases

aojea commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -644,7 +644,6 @@ func (kl *Kubelet) patchNodeStatus(originalNode, node *v1.Node) (*v1.Node, error
 		return nil, err
 	}
 	kl.lastStatusReportTime = kl.clock.Now()
-	kl.setLastObservedNodeAddresses(updatedNode.Status.Addresses)
```
@cartermckinnon this will also improve the latency on cloud providers, since it will not have to wait for the patch of the kubelet,

bart0sh commented:
/triage accepted

cartermckinnon commented:
Thanks @aojea!

/lgtm

aojea commented:
/test pull-kubernetes-integration


> {Failed  === RUN   TestKMSv2ProviderKeyIDStaleness/regular_gcm

liggitt commented:
> > {Failed  === RUN   TestKMSv2ProviderKeyIDStaleness/regular_gcm

cc @enj @aramase

enj commented:
> > > {Failed  === RUN   TestKMSv2ProviderKeyIDStaleness/regular_gcm
> 
> cc @enj @aramase

I see the bug (in the test), will sort it out soon.

aramase commented:
@aojea does this PR fix https://github.com/kubernetes/kubernetes/issues/130415?

aojea commented:
> @aojea does this PR fix #130415?

I don't have very clear the entire issue there, node readiness is more complex that cni only, what this does is to decouple the gathering of the node.addresses needed for the CSR generation from the complex loop that happens on nodeUpdateStatus and it returns the actual ips existing in the node.

Since nodeUpdateStatus runs periodically it guaranteed that the address will be eventually set, but as we found out, depending on that loop is very brittle and not really necessary, since the kubelet has access to the existing node object and can return the addresses directly

aojea commented:
/assign @SergeyKanzhelev @thockin

liggitt commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -466,3 +466,13 @@ func (kl *Kubelet) setCachedMachineInfo(info *cadvisorapiv1.MachineInfo) {
+func (kl *Kubelet) getLastObservedNodeAddresses() []v1.NodeAddress {
+	node, err := kl.GetNode()
+	if err != nil || node == nil {
+		klog.ErrorS(err, "fail to obtain node from local cache", "node", kl.nodeName)
```
I'm on the fence about logging an error here... on init while the node is getting its client cert and registering the node, this will be nil

aojea replied:
did I copied this from some place?

aojea replied:
let me add verbosity so it can be toggled

aojea replied:
updated to V(4)

aojea commented:
@liggitt pushed a change to increase the logging verbosity

https://github.com/kubernetes/kubernetes/compare/269cf8b38436ad2f467fd3101a018ab1b70466f8..b6334ae0542d74327d41538996d27b097630cac9

SergeyKanzhelev commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -466,3 +466,13 @@ func (kl *Kubelet) setCachedMachineInfo(info *cadvisorapiv1.MachineInfo) {
+func (kl *Kubelet) getLastObservedNodeAddresses() []v1.NodeAddress {
+	node, err := kl.GetNode()
+	if err != nil || node == nil {
+		klog.V(4).InfoS("fail to obtain node from local cache", "node", kl.nodeName)
```
can you please output the error here

SergeyKanzhelev reviewed (approved):
/lgtm

aojea commented:
this unit windows tests failing does not seem related but to require some attention
```
{Failed  === RUN   TestMemoryPressure_VerifyPodStatus
--- FAIL: TestMemoryPressure_VerifyPodStatus (0.03s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x30 pc=0x23dd824]

goroutine 50 [running]:
testing.tRunner.func1.2({0x25f6540, 0x3cb4ac0})
	C:/Program Files/Go/src/testing/testing.go:1632 +0x225
testing.tRunner.func1()
	C:/Program Files/Go/src/testing/testing.go:1635 +0x359
panic({0x25f6540?, 0x3cb4ac0?})
	C:/Program Files/Go/src/runtime/panic.go:785 +0x132
k8s.io/kubernetes/pkg/kubelet/eviction.makeMemoryAvailableSignalObservation(0xc0005de630)
	C:/kubernetes/pkg/kubelet/eviction/helpers_windows.go:35 +0x144
k8s.io/kubernetes/pkg/kubelet/eviction.makeSignalObservations(0xc0005de630)
	C:/kubernetes/pkg/kubelet/eviction/helpers.go:845 +0x1a5
k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).synchronize(0xc0005b1600, {0x2bf2a40, 0xc00040c650}, 0xc000063ba0)
	C:/kubernetes/pkg/kubelet/eviction/eviction_manager.go:306 +0x8d7
k8s.io/kubernetes/pkg/kubelet/eviction.TestMemoryPressure_VerifyPodStatus(0xc000356680)
	C:/kubernetes/pkg/kubelet/eviction/eviction_manager_test.go:322 +0x9ed
testing.tRunner(0xc000356680, 0x29e3df8)
	C:/Program Files/Go/src/testing/testing.go:1690 +0xcb
created by testing.(*T).Run in goroutine 1
	C:/Program Files/Go/src/testing/testing.go:1743 +0x377
```

tuminoid commented:
Uhh, it looks like the recently enabled Windows merge blocking job isn't very stable :(

aojea commented:
> @aojea: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-unit-windows-master	[174a9c8](https://github.com/kubernetes/kubernetes/commit/174a9c86528766bae7dbe9508c8886d5019dd70e)	[link](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130362/pull-kubernetes-unit-windows-master/1904083847360811008)	false	`/test pull-kubernetes-unit-windows-master`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=130362). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aaojea). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

indeed https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-unit-windows-master , I see it failing in a lot of PRs

/cc @BenTheElder @marosset

marosset commented:
> Uhh, it looks like the recently enabled Windows merge blocking job isn't very stable :(

That job isn't merge blocking and apologies for the failure, we've been putting a lot of effort into stabilizing the job in the v1.33 release

marosset commented:
/skip pull-kubernetes-unit-windows-master

BenTheElder commented:
> a rebase should fix that signal up (but feel free to merge without a rebase)

rebase shouldn't be necessary when testing in prow, we locally merge the PR as it would be if merged on github before running the tests (and record the commit)

just /retest will use the latest of (PR branch, Target branch)

BenTheElder commented:
> Ah, we don't do that with the Windows unit test jobs.
We just pull the PR changes and run the tests

We should be using the clone provided by prow, because otherwise we're recording the wrong metadata about what we tested, in periodic release signal AND in presubmit.

the commits are recorded and used by the merge robot and testgrid

aojea commented:
and also we are consistent, we should not have to rebase our PRs

marosset commented:
> should be using the clone provided by prow, because otherwise we're record

In order to run the Windows unit tests, we need to provision a Windows machine in an Azure sub hosted for kubernetes community work, and run the tests on that machine. 
If the commits are recorded and used by the merge bot and testgrid, we should be able to fetch them from the windows machine too, correct?

Looking at [podinfo.json](https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/130362/pull-kubernetes-unit-windows-master/1904221755271024640/podinfo.json), is `PULL_PULL_SHA` the SHA of the recorded merge commit?
If so we can try pulling that (right now we are pulling `PULL_NUMBER`

enj commented:
/lgtm cancel

To stop CI spam.

cartermckinnon commented:
/retest

cartermckinnon commented:
Might need a rebase to fix the tests?

aojea commented:
> Might need a rebase to fix the tests?

it shouldn't but it will retrigger the jobs, @cartermckinnon can you lgtm again?

aojea commented:
/retest

>    streamtranslator_test.go:249:  # HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
         # TYPE apiserver_stream_translator_requests_total counter
         apiserver_stream_translator_requests_total{code="200"} 1
        -apiserver_stream_translator_requests_total{code="400"} 1


the window unit test are panicing and look very unstable https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-unit-windows-master , they should not be informing in current state since create confusion

aojea commented:
> /lgtm cancel
> 
> To stop CI spam.

someone can lgtm again, the windows job does not seem related and is also not very stable

BenTheElder commented:
/skip

BenTheElder commented:
/lgtm
re-applying
